File: sdk/src/main/java/com/robotemi/sdk/TemiSdkServiceConnection.java
package com.robotemi.sdk;

import android.annotation.SuppressLint;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.UiThread;

import com.robotemi.sdk.constants.SdkConstants;

class TemiSdkServiceConnection {

    private static final String TAG = "TemiSdkServiceConnection";

    @NonNull
    private final ServiceConnection serviceConnection = new ServiceConnection() {

        @SuppressLint({"LogNotTimber", "LongLogTag"})
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            Log.d(TAG, "onServiceConnected(ComponentName, IBinder) (name=" + name + ", service=" + service + ", thread=" + Thread.currentThread().getName() + ")");
            final ISdkService sdkService = ISdkService.Stub.asInterface(service);
            Robot.Companion.getInstance().setSdkService(sdkService);
        }

        @SuppressLint({"LogNotTimber", "LongLogTag"})
        @Override
        public void onServiceDisconnected(ComponentName name) {
            Log.d(TAG, "onServiceDisconnected(ComponentName) (name=" + name + ", thread=" + Thread.currentThread().getName() + ")");
            forceStop();
        }
    };

    @SuppressLint({"LogNotTimber", "LongLogTag"})
    private static Intent getSdkServiceIntent(String appId) {
        Log.d(TAG, "getSdkServiceIntent()");
        final ComponentName componentName = new ComponentName(appId, BuildConfig.SDK_SERVICE_CLASS_NAME);
        final Intent intent = new Intent();
        intent.setComponent(componentName);
        return intent;
    }

    @SuppressLint({"LogNotTimber", "LongLogTag"})
    @UiThread
    void startConnection(@NonNull final Context context) {
        Log.d(TAG, "startConnection(Context)");
        if (context.bindService(getSdkServiceIntent(SdkConstants.TEMI_USA), serviceConnection, Context.BIND_AUTO_CREATE)) {
            Log.d(TAG, "bindServiceUsa=true");
        } else if (context.bindService(getSdkServiceIntent(SdkConstants.TEMI_CHINA), serviceConnection, Context.BIND_AUTO_CREATE)) {
            Log.d(TAG, "bindServiceChina=true");
        } else {
            Log.w(TAG, "bindService=false");
            forceStop();
        }
    }

    @SuppressLint({"LogNotTimber", "LongLogTag"})
    private void forceStop() {
        Log.d(TAG, "forceStop()");
        android.os.Process.killProcess(android.os.Process.myPid());
    }
}

File: sdk/src/main/java/com/robotemi/sdk/TemiSdkServiceConnection.kt
package com.robotemi.sdk

import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.os.IBinder
import android.os.Process
import android.util.Log

import androidx.annotation.NonNull
import androidx.annotation.UiThread

import com.robotemi.sdk.constants.SdkConstants

internal class TemiSdkServiceConnection {

    companion object {
        private const val TAG = "TemiSdkServiceConnection"

        @SuppressLint("LongLogTag")
        private fun getSdkServiceIntent(appId: String): Intent {
            Log.d(TAG, "getSdkServiceIntent()")
            val componentName = ComponentName(appId, BuildConfig.SDK_SERVICE_CLASS_NAME)
            return Intent().apply { component = componentName }
        }
    }

    @Volatile
    private var isLauncherExist = false

    @NonNull
    private val serviceConnection = object : ServiceConnection {

        @SuppressLint("LongLogTag")
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            Log.d(TAG, "onServiceConnected(ComponentName, IBinder) (name=$name, service=$service)")
            val sdkService = ISdkService.Stub.asInterface(service)
            Robot.getInstance().setSdkService(sdkService)
        }

        @SuppressLint("LongLogTag")
        override fun onServiceDisconnected(name: ComponentName) {
            Log.d(TAG, "onServiceDisconnected(ComponentName) (name=$name)")
            forceStop()
        }
    }

    @SuppressLint("LongLogTag")
    @UiThread
    internal fun startConnection(context: Context) {
        Log.d(TAG, "startConnection(Context)")
        isLauncherExist = isPackageExist(context, SdkConstants.TEMI_USA) ||
                isPackageExist(context, SdkConstants.TEMI_CHINA)
        when {
            context.bindService(
                getSdkServiceIntent(SdkConstants.TEMI_USA),
                serviceConnection,
                Context.BIND_AUTO_CREATE
            ) -> {
                Log.d(TAG, "bindServiceUsa=true")
            }
            context.bindService(
                getSdkServiceIntent(SdkConstants.TEMI_CHINA),
                serviceConnection,
                Context.BIND_AUTO_CREATE
            ) -> {
                Log.d(TAG, "bindServiceChina=true")
            }
            else -> {
                Log.w(TAG, "bindService=false")
                forceStop()
            }
        }
    }

    @SuppressLint("LongLogTag")
    private fun forceStop() {
        Log.d(TAG, "forceStop()")
        if (isLauncherExist) {
            Process.killProcess(Process.myPid())
        }
    }

    private fun isPackageExist(context: Context, targetPackage: String): Boolean {
        val pm: PackageManager = context.packageManager
        try {
            pm.getPackageInfo(targetPackage, 0)
        } catch (e: PackageManager.NameNotFoundException) {
            return false
        }
        return true
    }
}
